### PR TITLE
Add `vec_proxy()` and implement size and slice genericity with proxies

### LINKED ISSUE #124: Make vec_size() and vec_slice() generic
Since these expose the fundamental interface to vectors

### PR DESCRIPTION
**Pull request by lionel-**
Closes #124.

* New `vec_proxy()` generic. This should become the main customisation point in vctrs. It only needs to be implemented when the base type of a vector class is not its own data.

* Use `vec_proxy()` in `vec_slice()` and `vec_size()` to handle S3 vectors.

### CODE REVIEW

File: NEWS.md [hunk 1]
```diff
@@ -1,5 +1,10 @@
 # vctrs 0.1.0.9000
 
+* New `vec_proxy()` generic. This is the main customisation point in
```
Comment (hadley) [inline NEWS.md]: Do you have an example where someone would need to implement this? Like does it change anything in the S3 vector? Or is this primarily needed to allow vctrs to with existing vector classes? In that case, can you point to an example where you'd need to define a `vec_proxy()` method?

File: R/data.R [hunk 2]
```diff
@@ -1,11 +1,26 @@
...
-#' @param x A vector
+#' * `vec_data()` returns unstructured data. The only attributes
+#'   preserved are names.
+#'
+#'   Currently, due to the underlying memory architecture of R, this
+#'   creates a full copy of the data.
+#'
+#' * `vec_proxy()` may return structured data. This generic is the
+#'   main customisation point in vctrs, along with [vec_restore()].
+#'   You should only implement it when your type is not its own data,
```
Comment (hadley) [inline R/data.R]: I don't know what "your type is not its own data" means

File: NEWS.md (hunk 1 above)
Comment (hadley) [reply NEWS.md]: Should we implement `vec_proxy.POSIXlt()` as an example?

Comment (lionel-) [reply NEWS.md]: With the new scheme we discussed on Slack (records return data frames), `vec_proxy.POSIXlt()` now returns a data frame.

I have updated the documentation in NEWS and `?vec_proxy`.

**Comment by lionel-** [comment]
* The default `vec_proxy()` method now returns its input.

* Record types are proxied by data frames.

* `vec_is_vector()` is now unexported as it no longer needs to be a customisation point. To find out whether a type is a vector or a scalar, we just check that the proxy of an object is an atomic, bare list, or data frame.

  Consequently, S3 lists must implement `vec_proxy()` to unclass themselves or return a data frame.

File: R/data.R [hunk 3]
```diff
@@ -1,16 +1,71 @@
...
+#' @description
+#'
 #' Extract the data underlying an S3 vector object, i.e. the underlying
 #' (named) atomic vector or list.
-#' Currently, due to the underlying memory architecture
-#' of R, this creates a full copy of the data.
 #'
-#' @param x A vector
-#' @export
+#' * `vec_data()` returns unstructured data. The only attributes
```
Comment (hadley) [inline R/data.R]: And dims, and dimnames?

**Comment by lionel-** [comment]
I was wondering if `vec_proxy_equal()` could be entirely removed, now that we have `vec_proxy()`.

**Comment by hadley** [comment]
It would require some rewriting of equality comparison, because `vec_proxy_equality()` is recursive

**Comment by lionel-** [comment]
Yup and this would be congruent with how I suggested to rewrite `vec_equal()` a while ago.

**Pull request merged**